core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -391,17 +401,34 @@ public TableMetadata addStagedSnapshot(Snapshot snapshot) {
   }
 
   public TableMetadata replaceCurrentSnapshot(Snapshot snapshot) {
-    List<Snapshot> newSnapshots = ImmutableList.<Snapshot>builder()
-        .addAll(snapshots)
-        .add(snapshot)
-        .build();
-    List<HistoryEntry> newSnapshotLog = ImmutableList.<HistoryEntry>builder()
-        .addAll(snapshotLog)
-        .add(new SnapshotLogEntry(snapshot.timestampMillis(), snapshot.snapshotId()))
-        .build();
-    return new TableMetadata(null, uuid, location,
-        snapshot.timestampMillis(), lastColumnId, schema, defaultSpecId, specs, properties,
-        snapshot.snapshotId(), newSnapshots, newSnapshotLog, addPreviousFile(file, lastUpdatedMillis));
+    // there can be operations (viz. rollback, cherrypick) where an existing snapshot could be replacing current
```

Comment:
This method isn't used for rollback; that calls `rollbackTo`. I think it would be easier to update `rollbackTo` to be `setCurrentSnapshotTo`. Then cherrypick should use either `replaceCurrentSnapshot` if it is creating a new snapshot, or `setCurrentSnapshotTo` if it is using an existing snapshot (the fast-forward case).

The reason I'd structure it this way is to keep the methods in this file independent of the operations that they are used for. That way `replaceCurrentSnapshot` doesn't care how it is called and always does the same thing: adds a snapshot, updates the current pointer, and updates the snapshot log. Similarly, `setCurrentSnapshotTo` will have simple behavior.

Reply:
The only issue with keeping `replaceCurrentSnapshot` and `setCurrentSnapshotTo` separate is that cherrypick and rollback workflows will then need to override `commit()` and hence re-implement retry logic inside their `commit()` logic (all the things `SnapshotProducer` is doing for us today).  The reason I incorporated logic for existing and new snapshot inside `replaceCurrentSnapshot` was coz now rollback and cherrypick workflows rely on this method ( and not `setCurrentSnapshotTo` - erstwhile `rollbackTo` ).

Reply:
The way i see this is,  TableMetadata still doesn't need to know about rollback or cherrypick but can handle replacing `current_snapshot` with an existing snapshot OR a new snapshot. It's just handling both cases. viz.
```
setCurrentSnapshotTo(snapshot) {
    // updates the current pointer, 
    // and updates the snapshot log.
}

replaceCurrentSnapshot(snapshot) {
  if (exists) {
    setCurrentSnapshotTo(snapshot)
  } else {
     // adds a snapshot, 
    // updates the current pointer, 
    // and updates the snapshot log.
  }
}
```

Reply:
Also, now that the deprecated Rollback api's impl also translates to ManageSnaphots api.. that too doesn't implement commit(), ergo doesn't have to call `setCurrentSnapshotTo()` (it earlier used  - `ops.commit(base.rollbackTo())` ) .

Reply:
Sounds good to me. I agree that it's worth it to avoid overriding commit. Should we also make `setCurrentSnapshotTo` private?